api/types/saml_idp_service_provider.go

```diff
@@ -143,3 +144,331 @@ func (s SAMLIdPServiceProviders) Less(i, j int) bool { return s[i].GetName() < s
 
 // Swap swaps two service providers.
 func (s SAMLIdPServiceProviders) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+// AppServerOrSAMLIdPServiceProvider holds either an AppServer or a SAMLIdPServiceProvider resource (never both).
+// This is the resource used for WebUI requests to list Applications since we want to list both AppServers and
+// SAMLIdPServiceProviders in the UI.
+type AppServerOrSAMLIdPServiceProvider interface {
+	ResourceWithLabels
+	GetAppServer() *AppServerV3
+	GetSAMLIdPServiceProvider() *SAMLIdPServiceProviderV1
+	GetAppOrServiceProviderName() string
+	GetAppOrServiceProviderDescription() string
+	GetAppOrServiceProviderPublicAddr() string
+	IsAppServer() bool
+}
+
+const (
+	// This is the `Description` of a SAML IdP Service Provider to show when listing it in the WebUI.
+	SAMLIdPServiceProviderDescription = "SAML Application"
+)
+
+// // GetAppServer returns the AppServer in this AppServerOrSAMLIdPServiceProvider.
+// func (a *AppServerOrSAMLIdPServiceProviderV1) GetAppServer() AppServer {
+// 	appOrSP := a.AppServerOrSAMLIdPServiceProvider
+// 	return appOrSP.(AppServer)
+// }
+
+// // GetAppServer returns the GetSAMLIdPServiceProvider in this AppServerOrSAMLIdPServiceProvider.
+// func (a *AppServerOrSAMLIdPServiceProviderV1) GetSAMLIdPServiceProvider() SAMLIdPServiceProvider {
+// 	appOrSP := a.AppServerOrSAMLIdPServiceProvider
+// 	return appOrSP.(SAMLIdPServiceProvider)
+// }
+
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetKind() string {
+	if a.IsAppServer() {
+		return KindSAMLIdPServiceProvider
+	}
+	return KindAppServer
+}
+
+// GetAppOrServiceProviderName returns the name of either the App or the SAMLIdPServiceProvider, depending on which one
+// the AppServerOrSAMLIdPServiceProvider holds.
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetAppOrServiceProviderName() string {
+	if a.IsAppServer() {
+		return a.GetAppServer().GetApp().GetName()
+	}
+	return a.GetSAMLIdPServiceProvider().GetName()
+}
+
+// GetAppOrServiceProviderDescription returns the name of either the App or the SAMLIdPServiceProvider, depending on which one
+// the AppServerOrSAMLIdPServiceProvider holds.
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetAppOrServiceProviderDescription() string {
```

Comment:
do these func names have to be very specific? 
```suggestion
func (a *AppServerOrSAMLIdPServiceProviderV1) GetDescription() string {
```